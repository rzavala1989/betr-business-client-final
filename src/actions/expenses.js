// import uuid from 'uuid';  /* <---NOT NEEDED */

import {API_BASE_URL} from "../config";
import {normalizeResponseErrors} from './utils';

export const FETCH_EXPENSE_DATA_SUCCESS = "FETCH_EXPENSE_DATA_SUCCESS";
export const fetchExpenseDataSuccess = data => ({
  type: FETCH_EXPENSE_DATA_SUCCESS,
  data
});

export const FETCH_EXPENSE_DATA_ERROR = "FETCH_EXPENSE_DATA_ERROR";
export const fetchExpenseDataError = error => ({
  type: FETCH_EXPENSE_DATA_ERROR,
  error
});

export const fetchExpenseData = () => (dispatch, getState) => {
  const authToken = getState().auth.authToken;
  return fetch(`${API_BASE_URL}/expenses`, {
    method: "GET",
    headers: {
      // Provide our auth token as credentials
      Authorization: `Bearer ${authToken}`
    }
  })
    .then(res => normalizeResponseErrors(res))
    .then(res => res.json())
    .then(data => dispatch(fetchExpenseDataSuccess(data)))
    .catch(err => {
      dispatch(fetchExpenseDataError(err));
    });
};

export const POST_EXPENSE_DATA_SUCCESS = "POST_EXPENSE_DATA_SUCCESS";
export const postExpenseDataSuccess = data => ({
  type: POST_EXPENSE_DATA_SUCCESS,
  data
});

export const POST_EXPENSE_DATA_ERROR = "POST_EXPENSE_DATA_ERROR";
export const postExpenseDataError = error => ({
  type: POST_EXPENSE_DATA_ERROR,
  error
});

export const postExpense = expense => (dispatch, getState) => {
  const authToken = getState().auth.authToken;
  return fetch(`${API_BASE_URL}/expenses`, {
    method: "POST",
    headers: {
      // Provide our auth token as credentials
      Authorization: `Bearer ${authToken}`,
      "Content-type": "application/json"
    },
    body: JSON.stringify(expense)
  })
    .then(res => normalizeResponseErrors(res))
    .then(res => res.json())
    .then(data => dispatch(postExpenseDataSuccess(data)))
    .catch(err => {
      dispatch(postExpenseDataError(err));
    });
};

export const EDIT_EXPENSE_SUCCESS = "EDIT_EXPENSE_SUCCESS";
export const editExpenseSuccess = (id, updates) => ({
  type: EDIT_EXPENSE_SUCCESS,
  id,
  updates
});

export const EDIT_EXPENSE_ERROR = "EDIT_EXPENSE_ERROR";
export const editExpenseError = error => ({
  type: EDIT_EXPENSE_ERROR,
  error
});

export const editExpense = (id, updates) => (dispatch, getState) => {
  updates.id = id;
  const authToken = getState().auth.authToken;
  fetch(`${API_BASE_URL}/expenses/${id}`, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${authToken}`
    },
    body: JSON.stringify(updates)
  })
    .then(res => dispatch(editExpenseSuccess(updates)))
    .catch(err => {
      dispatch(editExpenseError(err));
    });
};

export const DELETE_EXPENSE_SUCCESS =
  "DELETE_EXPENSE_SUCCESS";
export const deleteExpenseSuccess = ({id}) => ({
  type: DELETE_EXPENSE_SUCCESS,
  id
});

export const DELETE_EXPENSE_ERROR = "DELETE_EXPENSE_ERROR";
export const deleteExpenseError = error => ({
  type: DELETE_EXPENSE_ERROR,
  error
});

export const deleteExpense = ({id}) => (dispatch, getState) => {
  const authToken = getState().auth.authToken;
  return fetch(`${API_BASE_URL}/expenses/${id}`, {
    method: "DELETE",
    headers: {
      // Provide our auth token as credentials
      Authorization: `Bearer ${authToken}`
    }
  }).then(res => dispatch(deleteExpenseSuccess({id})));
};

// export const SET_SELECTED_EXPENSE = "SET_SELECTED_EXPENSE";
// export const setSelectedExpense = (expense, username) => ({
//   type: SET_SELECTED_EXPENSE,
//   expense,
//   username
// });

// ________________________________

// ADD_EXPENSE
// export const addExpense = (
//   {
//     description = '',
//     note = '',
//     amount = 0,
//     createdAt = 0
//   } = {}
// ) => ({
//   type: 'ADD_EXPENSE',
//   expense: {
//     description,
//     note,
//     amount,
//     createdAt
//   }
// });

// REMOVE_EXPENSE
// export const removeExpense = ({ id } = {}) => ({
//   type: 'REMOVE_EXPENSE',
//   id
// });

// EDIT_EXPENSE
// export const editExpense = (id, updates) => ({
//   type: 'EDIT_EXPENSE',
//   id,
//   updates
// });
