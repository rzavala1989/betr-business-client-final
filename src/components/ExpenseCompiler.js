import React from 'react';
import {connect} from 'react-redux';
import {Link} from 'react-router-dom';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import selectTotal from '../selectors/total-expenses';

export const ExpenseCompiler = ({ expenseCount, expensesTotal }) => {
  const numberOfWords = expenseCount === 1 ? 'expense' : 'expenses';
  const formattedTotal = numeral(expensesTotal/100).format('$0,0.00');
  if (expenseCount === 0) {
    return (
      <div className="page-header">
        <div className="container">
          <h1 className="page-header__title">You have no expenses! Spend some money</h1>
          <div className="page-header__add">
            <Link className="button button-db" to="/create">Add Business Expense</Link>
          </div>
        </div>
      </div>
    );
  } else {
    return (
      <div className="page-header">
        <div className="container">
          <h1 className="page-header__title"><span>{expenseCount}</span> {numberOfWords} totalling <span>{formattedTotal}</span></h1>
          <div className="page-header__add">
            <Link className="button" to="/create">Add Another Expense</Link>
          </div>
        </div>
      </div>
    );
  }
};

const mapStateToProps = (state) => {
  const visibleExpenses = selectExpenses(state.expenses.expenses, state.filters);
  return {
    expenseCount: visibleExpenses.length,
    expensesTotal: selectTotal(visibleExpenses)
  };
};

export default connect(mapStateToProps)(ExpenseCompiler);
