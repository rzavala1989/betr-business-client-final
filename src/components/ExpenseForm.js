import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';


export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: ''
    };
  }
  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  onAmountChange = (e) => {
    const amount = e.target.value;
    //conditional logic to determine correct format for currency
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };
  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calenderFocused: focused }));
  };
  onSubmit = (e) => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: "Please fill both Description and Amount fields." })); 
    } else {
      this.setState(() => ({ error: "" }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };
  render() {
    return (
        <form className="form" onSubmit={this.onSubmit}>
          {this.state.error && <p className="form__error">{this.state.error}</p>}
          <select
            className="input--text"
            value={this.state.description}
            onChange={this.onDescriptionChange}
            autoFocus
            required
          >
            <option value="">Please select an option</option>
            <option value="Travel">Travel</option>
            <option value="Entertainment">Entertainment</option>
            <option value="Internet and Phone Bills">Internet and Phone Bills</option>
            <option value="Health Insurance Premium">Health Insurance Premium</option>
            <option value="Meals">Meals</option>
            <option value="Vehicle Use">Vehicle Use</option>
            <option value="Education Expenses">Education Expenses</option>
            <option value="Business Insurance">Business Insurance</option>
            <option value="Rent">Rent</option>
            <option value="Advertising">Advertising</option>
            <option value="Other">Other (Please elaborate in detail in notes)</option>
          </select>
          <input
            type="text"
            className="input--text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calenderFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            className="input--textarea"
            placeholder="Add a note for your expense (optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
          >
          </textarea>
          <div>
            <button className="button--add">Save Expense</button>
          </div>
        </form>
    )
  }
}