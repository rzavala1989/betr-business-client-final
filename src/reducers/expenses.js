// Expenses Reducer
import {
  FETCH_EXPENSE_DATA_SUCCESS,
  FETCH_EXPENSE_DATA_ERROR,
  POST_EXPENSE_DATA_SUCCESS,
  POST_EXPENSE_DATA_ERROR,
  EDIT_EXPENSE_SUCCESS,
  EDIT_EXPENSE_ERROR,
  DELETE_EXPENSE_SUCCESS,
  DELETE_EXPENSE_ERROR
} from "../actions/expenses";

const initialState = {
  expenses: []
};

export default function reducer(state = initialState, action) {
  if (action.type === FETCH_EXPENSE_DATA_SUCCESS) {
    return Object.assign({}, state, { expenses: action.data, error: null });
  }
  if (action.type === FETCH_EXPENSE_DATA_ERROR) {
    return Object.assign({}, state, { error: action.error });
  }
  if (action.type === POST_EXPENSE_DATA_SUCCESS) {
    return Object.assign({}, state, {
      expenses: [...state.expenses, action.data],
      error: null
    });
  }
  if (action.type === POST_EXPENSE_DATA_ERROR) {
    return Object.assign({}, state, { error: action.error });
  }
  if (action.type === EDIT_EXPENSE_SUCCESS) {
    return Object.assign({}, state, {
      expenses: state.expenses.map(
        expense =>
          expense.id === action.id ? action.updates: expense
      )
    });
  }
  if (action.type === EDIT_EXPENSE_ERROR) {
    return Object.assign({}, state, { expenses: action.error });
  }
  if (action.type === DELETE_EXPENSE_SUCCESS) {
    return Object.assign({}, state, {
      ...state.expenses.slice(0, action.index),
      ...state.expenses.slice(action.index + 1)
    });
  }
  if (action.type === DELETE_EXPENSE_ERROR) {
    return Object.assign({}, state, { error: action.error });
  }
  return state;
}

