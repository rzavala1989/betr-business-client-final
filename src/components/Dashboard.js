import React from 'react';
import {connect} from 'react-redux';
import requiresLogin from './Requires-login';
import {fetchExpenseData} from '../actions/expenses';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpenseCompiler from "./ExpenseCompiler";

export class Dashboard extends React.Component {
    componentDidMount() {
        this.props.dispatch(fetchExpenseData());
    }

    render() {
        return (
            <div className="dashboard">
            <div className="bg-dashboard"></div>
                <div className="dashboard-anim">
                    <div className="inline-block dashboard-name-1">Hello {this.props.name}!</div>
                    <div className="inline-block dashboard-name-2"><span>Let's do Betr Business</span></div>
                </div>
                <ExpenseCompiler/>
                <ExpenseListFilters />
                <ExpenseList />
            </div>
        );
    }
}

const mapStateToProps = state => {
    const {currentUser} = state.auth;
    return {
        username: state.auth.currentUser.username,
        name: `${currentUser.firstName} ${currentUser.lastName}`,
        protectedData: state.protectedData.data
    };
};

export default requiresLogin()(connect(mapStateToProps)(Dashboard));
