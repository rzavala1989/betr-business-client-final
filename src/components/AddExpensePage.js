import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import{ postExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    console.log(expense);
    this.props.postExpense(expense);
    this.props.history.push('/');
  }
  render() {
    return (
      <div className="add-expense-page">
        <div className="bg-add-expense-page"></div>
        <div className="page-header">
          <div className="container">
            <h1 className="page-header__title">Add an Expense</h1>
          </div>
        </div>
        <div className="container">
          <ExpenseForm
            onSubmit={this.onSubmit}
          />
        </div>
      </div>
    )
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    postExpense: (expense) => dispatch(postExpense(expense))
  };
};

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
