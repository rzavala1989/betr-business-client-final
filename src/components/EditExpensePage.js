import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, deleteExpense } from '../actions/expenses';


export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    console.log(expense)
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };
  onRemove = () => {
    this.props.deleteExpense({ id: this.props.expense.id });
    this.props.history.push('/');
  };
  render() {
    return (
      <div className="edit-expense-page">
      <div className="bg-edit-expense-page"></div>
        <div className="page-header">
          <div className="container">
            <h1 className="page-header__title">Edit Your Expense</h1>
          </div>
        </div>
        <div className="container">
          <ExpenseForm
            expense={this.props.expense}
            onSubmit={this.onSubmit}
          />
          <button className="button--remove" onClick={this.onRemove}>Remove Expense Forever</button>
        </div>
      </div>
    );
  }
};

const mapStateToProps = (state, props) => ({
  expense: state.expenses.expenses.find((expense) => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
  editExpense: (id, expense) => dispatch(editExpense(id, expense)),
  deleteExpense: (data) => dispatch(deleteExpense(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);







// const EditExpensePage = (props) => {
//   return (
//     <div className="edit-expense-page">
//       <div className="bg-edit-expense-page"></div>
//         <div className="page-header">
//           <div className="container">
//             <h1 className="page-header__title">Edit Your Expense</h1>
//           </div>
//         </div>
//         <div className="container">
//           <ExpenseForm
//             expense={props.expense}
//             onSubmit={(expense) => {
//               props.dispatch(editExpense(props.expense.id, expense));
//               props.history.push('/');
//             }}
//           />
//           <button 
//             className="button--remove" 
//             onClick={() => {
//               props.dispatch(deleteExpense({ id: props.expense.id }));
//               props.history.push('/');
//             }}>Remove Expense Forever</button>
//         </div>
//       </div>
//   )
// };

// const mapStateToProps = (state, props) => ({
//   expense: state.expenses.expenses.find((expense) => expense.id === props.match.params.id)
// });

// export default connect(mapStateToProps)(EditExpensePage);


