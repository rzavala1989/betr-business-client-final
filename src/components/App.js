import React from 'react';
import {connect} from 'react-redux';
import {Route, withRouter, Switch} from 'react-router-dom';

import HeaderBar from './Header-bar';
import LandingPage from './Landing-page';
import Dashboard from './Dashboard';
import AddExpensePage from './AddExpensePage';
import EditExpensePage from './EditExpensePage';
import HelpPage from './HelpPage';
import {Footer} from './Footer';
import NotFoundPage from './NotFoundPage';
import RegistrationPage from './Registration-page';
import {refreshAuthToken} from '../actions/auth';

export class App extends React.Component {
    componentDidUpdate(prevProps) {
        if (!prevProps.loggedIn && this.props.loggedIn) {
            // When we are logged in, refresh the auth token periodically
            this.startPeriodicRefresh();
        } else if (prevProps.loggedIn && !this.props.loggedIn) {
            // Stop refreshing when we log out
            this.stopPeriodicRefresh();
        }
    }

    componentWillUnmount() {
        this.stopPeriodicRefresh();
    }

    startPeriodicRefresh() {
        this.refreshInterval = setInterval(
            () => this.props.dispatch(refreshAuthToken()),
            60 * 60 * 1000 // One hour
        );
    }

    stopPeriodicRefresh() {
        if (!this.refreshInterval) {
            return;
        }

        clearInterval(this.refreshInterval);
    }

    render() {
        return (
            <div className="app">
                <HeaderBar />
                <Switch>
                    <Route exact path="/" component={LandingPage} />
                    <Route path="/dashboard" component={Dashboard} />
                    <Route path="/create" component={AddExpensePage} />
                    <Route path="/edit/:id" component={EditExpensePage} />
                    <Route path="/help" component={HelpPage} />
                    <Route path="/register" component={RegistrationPage} />
                    <Route component={NotFoundPage} />
                </Switch>
                <Footer/>
            </div>
        );
    }
}

const mapStateToProps = state => ({
    hasAuthToken: state.auth.authToken !== null,
    loggedIn: state.auth.currentUser !== null
});

// Deal with update blocking - https://reacttraining.com/react-router/web/guides/dealing-with-update-blocking
export default withRouter(connect(mapStateToProps)(App));
