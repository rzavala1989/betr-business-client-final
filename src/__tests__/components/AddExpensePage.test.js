import React from 'react';
import { shallow } from 'enzyme';
import {AddExpensePage} from '../../components/AddExpensePage';
import expenses from '../../test-data/expenses';
import {Provider} from 'react-redux';

let addExpense, history, wrapper;

beforeEach(() => {
  addExpense = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(<Provider><AddExpensePage history={history} /></Provider>);
});

test('should render AddExpensePage correctly', () => {
  expect(wrapper).toMatchSnapshot();
});


