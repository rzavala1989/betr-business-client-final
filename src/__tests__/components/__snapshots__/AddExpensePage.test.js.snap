// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddExpensePage correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Provider>
    <AddExpensePage
        history={
            Object {
                "push": [Function],
              }
        }
    />
</Provider>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "history": Object {
        "push": [Function],
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "history": Object {
          "push": [Function],
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): Object {
    "store": undefined,
    "storeSubscription": null,
  },
}
`;
